rclcpp/include/rclcpp/qos_event.hpp

```diff
@@ -131,21 +133,33 @@ class QOSEventHandler : public QOSEventHandlerBase
     }
   }
 
-  /// Execute any entities of the Waitable that are ready.
-  void
-  execute() override
+  /// Take data so that the callback cannot be scheduled again
+  std::shared_ptr<void>
+  take_data() override
   {
     EventCallbackInfoT callback_info;
-
     rcl_ret_t ret = rcl_take_event(&event_handle_, &callback_info);
     if (ret != RCL_RET_OK) {
       RCUTILS_LOG_ERROR_NAMED(
         "rclcpp",
         "Couldn't take event info: %s", rcl_get_error_string().str);
-      return;
+      return nullptr;
     }
+    return std::static_pointer_cast<void>(
+      std::make_shared<EventCallbackInfoT>(
+        callback_info));
```

Comment:
nitpick: unnecessary new line here

Reply:
Done in https://github.com/ros2/rclcpp/pull/1241/commits/4d21b6eb6404b32b687303915c9c5b36e73812e9.